ubclaunchpad/money-manager
# Auth middleware mathew

## Description ✍️

- Setup FastAPI middleware to authenticate user based on cookie auth header.
- It should make sure the user is logged in with the correct email and password and if not return an unauthorized error.
- If the user is logged in, then add the user email along with the rest of the request.

### Changes ⚙️

- added `hashed_password`,`active` fields to user model
- added `passlib` ,` python-jose` and `bcrypt` to requirements.txt
- Implemented a login flow on the backend 
- added password hashing: BE receives plain password, hashes it and compares it to stored `hashed_password` in the database
- Added tokens(JWT tokens) : after a user logs in, create a login token with an expiration time (currently 30 minutes but easily changeable using a parameter in the code), if a user comes back before the expiration time, they will still be logged in. User has to sign in again after expiration of token
- Just made sure it was working, will clean up code later

## How to test authentication
1. go to `localhost:8000/docs`
2. try sending GET request using `users/me/`, should get `not authenticated`
3. click on green authorize button
4. login using these credentials, email: "testuser@moneymanager.com", password: "abc123"
5. try sending the get request again to /users/me. You should get a json object containing all data about the logged in user
6. Try logging in with an incorrect password, you should get a `401 unauthorized error`. Incorrect email should give you a `404`

## Checklist 🗒

- [x] Ran `black .` to format code in `./server`
- [x] Ran `npm run lint:fix` to format code in `./client`
- [x] Assigned PR to all authors
- [x] Requested at least one reviewer
- [x] Linked the PR to its respective issue in ZenHub
- [x] Replaced the X below with the issue number

Closes #38
